pep-0484.txt

```diff
@@ -276,6 +276,33 @@ implemented by deferring to ``isinstance(x, collections.abc.Callable)``.
 However, ``isinstance(x, typing.Callable[...])`` is not supported.
 
 
+Decorators
+----------
+
+Decorators can modify the types of the functions or classes they
+decorate. To annotate a decorated declaration with the eventual
+decorated type, put a type comment on the line with the decorator::
+
+  from typing import ContextManager, Generator
+  from contextlib import contextmanager
+
+  class DatabaseSession: ...
+
+  @contextmanager  # type: Callable[[str], ContextManager[DatabaseSession]]
+  def session(url: str) -> Generator[DatabaseSession, None, None]:
+      s = DatabaseSession(url)
+      try:
+          yield s
+      finally:
+          s.close()
+
+If both the decorator and the function being decorated have type
```

Comment:
@sixolet Could you clarify this paragraph a bit? It's not clear if "decorator" refers to the decorated declaration or the decorator function itself.

If "decorator" here is the decorated declaration (`@contextmanager`), then its type isn't really related to the type of the function being decorated. In the example above there is no point in matching `Callable[[str], Generator[DatabaseSession, None, None]` against `Callable[[str], ContextManager[DatabaseSession]]`.

Alternatively, if "decorator" here means the decorator function itself (`def contextmanager(...): ...`, then this paragraph is not related to the idea of adding type hints for decorated declarations.